stake-pool/stake-pool-test/tests/helpers/mod.rs

```diff
@@ -0,0 +1,332 @@
+
+        match self {
+            Env::SinglePool(_) => {
+                program_test.add_program(
+                    "spl_stake_birdbath",
+                    spool::id(),
+                    processor!(spool::processor::Processor::process),
+                );
+                program_test.deactivate_feature(stake_raise_minimum_delegation_to_1_sol::id());
+            }
+            Env::MultiPool(_) => {
+                program_test.add_program(
+                    "spl_stake_pool",
+                    mpool::id(),
+                    processor!(mpool::processor::Processor::process),
+                );
+                program_test.add_program(
+                    "spl_token_2022",
+                    spl_token_2022::id(),
+                    processor!(spl_token_2022::processor::Processor::process),
+                );
+            }
+        }
+
+        program_test.prefer_bpf(false);
+        program_test
+    }
+
+    /* XXX dunno if i need this
+        pub fn is_multi(&self) -> bool {
+            match self {
+                Env::SinglePool => false,
+                _ => true,
+            }
+        }
+    */
+
+    pub fn set_deposit_authority(&mut self, stake_deposit_authority: Keypair) {
+        match self {
+            Env::SinglePool(_) => panic!("dont do that"),
+            // TODO FIXME check that this actually works, clippy said i dont need to borrow...
+            Env::MultiPool(accounts) => {
+                accounts.stake_deposit_authority = stake_deposit_authority.pubkey();
+                accounts.stake_deposit_authority_keypair = Some(stake_deposit_authority);
+            }
+        }
+    }
+
+    pub fn set_reserve_lamports(&mut self, reserve_lamports: u64) {
+        match self {
+            Env::SinglePool(_) => panic!("dont do that"),
+            // TODO FIXME check that this actually works, clippy said i dont need to borrow...
+            Env::MultiPool(accounts) => {
+                accounts.reserve_lamports = reserve_lamports;
+            }
+        }
+    }
+
+    pub async fn initialize(
+        &self,
+        banks_client: &mut BanksClient,
+        payer: &Keypair,
+        recent_blockhash: &Hash,
+    ) -> Result<(), TransportError> {
+        match self {
+            Env::SinglePool(accounts) => {
+                accounts
+                    .initialize(banks_client, payer, recent_blockhash)
+                    .await
+            }
+            Env::MultiPool(accounts) => {
+                accounts
+                    .initialize(banks_client, payer, recent_blockhash)
+                    .await
+            }
+        }
+    }
+}
+
+// XXX ok im confused about parametrization again
+// * if i do an Accounts trait, and non-generic Env, i impl initialize on the trait
+//   env only exists to produce the accounts
+//   but because i dont return a concrete type i need to work in separate branch arms in tests?
```

Comment:
That depends. You can either use `dyn MyTraitType`, or an enum-with-value that does some form of `enum_dispatch` in its implementation, which is what `Env` is doing now. Since we don't need to worry about performance, it might simpler to use `Box<dyn MyTraitType>` (or Arc).

I'm probably missing some bits, but here's a potential approach. You define a trait with all of the functionality required for single pools only, ie. deposit, withdraw, initialize, create metadata, update metadata, along with functions to give you the important addresses (mint, stake account, etc). You have two concrete types, `SinglePool` and `MultiPool`, which each implement that minimal interface.

For tests on both single and multipool, you run everything through the trait. Since you have functions that give you the addresses, you should (?) be able to check through that.

For tests on just one or the other, you can directly return the concrete type from your builder, or expose `unwrap_multi()` / `unwrap_single()` functions to pull out the concrete type, and just work on that. That way you don't need to wrap tests in an `if let Env::(...)`